android screen orientation: 1st pass

quinlanj opened the pull request:
# Why

- first pass at screen orientation API for android, documentation here: https://github.com/expo/universe/pull/3280
- contains android impl, plus test-suite coverage for all the methods
- In addition to whats in the documentation PR, i also made the following changes:
-- Added listener removal methods `removeOrientationChangeListeners()` and `removeOrientationChangeListener(subscription)`. RN's NativeEventEmitters does something similar.
-- Added a way to retrieve a stringified native orientation lock via `getOrientationLockPlatformAsync()`. I figured if we have a platform setter `lockPlatformAsync`, we should have a platform getter.
-- Default orientation lock on Android is [UNSPECIFIED](https://developer.android.com/reference/android/content/pm/ActivityInfo.html#SCREEN_ORIENTATION_UNSPECIFIED) because this is what I've seen as default when testing on the Pixel and Nexus. On iOS I'm thinking of making the default policy `ALL_BUT_PORTRAIT_UPSIDE_DOWN`.

Related PRS:
Docs - https://github.com/expo/universe/pull/3280
JS - https://github.com/expo/expo/pull/3103
iOS - https://github.com/expo/expo/pull/3104
Android - https://github.com/expo/expo/pull/3105
Tests - https://github.com/expo/expo/pull/3106
Previous Feedback (1st pass android) - https://github.com/expo/expo/pull/2832

quinlanj committed:
android screen orientation: 1st pass

quinlanj committed:
android stuff

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ScreenOrientationModule.java:
```diff
@@ -55,47 +60,212 @@ public void onCatalystInstanceDestroy() {
+  static final String ERR_SCREEN_ORIENTATION_INVALID_ORIENTATION_LOCK = "ERR_SCREEN_ORIENTATION_INVALID_ORIENTATION_LOCK";
+  static final String ERR_SCREEN_ORIENTATION_INVALID_ORIENTATION_LOCK_MSG = "an invalid OrientationLock was passed in";
+
+  static final String UNKNOWN = "UNKNOWN";
```
Might make sense to define a Java enum for these (and give the enum instances a `toString()` method that returns a string value)

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ScreenOrientationModule.java:
```diff
@@ -55,47 +60,212 @@ public void onCatalystInstanceDestroy() {
+          orientation = LANDSCAPE_RIGHT;
+          break;
+        default:
+          throw new IllegalStateException("Unknown screen orientation.");
```
Seems like returning UNKNOWN would be simpler here and for the sole caller that converts the caught exception to UNKNOWN anyway.

ide commented on apps/test-suite/tests/ScreenOrientation.js:
```diff
@@ -0,0 +1,296 @@
+export const name = 'ScreenOrientation';
+
+// run callback every interval until it returns something truthy
+const pollUntilSuccessAsync = async (callback, interval = 500) => {
```
IMO it'd be preferable for implement this without polling and instead rely on events. The polling doesn't timeout and if tests are killed by Jasmine's timeout, this polling code will still run in the background indefinitely. (Perhaps we may want a small number of tests with safe polling in case the events break, but polling is always either too early or too late.) 

Instead of polling, we could structure this like:

```
// set up listener
add listener {
  if event.orientation !== desiredOrientation
    return
  if desiredOrientationLock !== null && event.orientationLock !== desiredOrientationLock
    return
  // yay, satisified
  remove listener
  cancel timeout
  resolve promise
}

// check current value
currentOrientation, currentOrientationLock = await [getOrientation(), getOrientationLock()]
  if currentOrientation !== desiredOrientation
    not satisfied
  if desiredOrientationLock !== null && currentOrientationLock !== desiredOrientationLock
    not satisfied
  // if we are satisfied
  remove listener
  resolve promise

// set up timeout
setTimeout {
  remove listener
  reject promise
}
```

ide commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ScreenOrientationModule.java:
```diff
@@ -55,47 +60,212 @@ public void onCatalystInstanceDestroy() {
+  public void lockAsync(String orientationLock, Promise promise) {
     Activity activity = getCurrentActivity();
     if (activity == null) {
+      promise.reject(ANDROID_NO_ACTIVITY, ANDROID_NO_ACTIVITY_MSG);
```
Do we need this?

esamelson reviewed (commented):
a few preliminary comments on top of James's

ide committed:
Update packages/expo/src/ScreenOrientation.ts

Co-Authored-By: quinlanj <quin@expo.io>

ide committed:
Update packages/expo/src/ScreenOrientation.ts

Co-Authored-By: quinlanj <quin@expo.io>

ide committed:
Update packages/expo/src/ScreenOrientation.ts

Co-Authored-By: quinlanj <quin@expo.io>

ide committed:
Apply suggestions from code review

Co-Authored-By: quinlanj <quin@expo.io>

ide committed:
Apply suggestions from code review

Co-Authored-By: quinlanj <quin@expo.io>

quinlanj commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ScreenOrientationModule.java:
```diff
@@ -55,47 +60,212 @@ public void onCatalystInstanceDestroy() {
+          orientation = LANDSCAPE_RIGHT;
+          break;
+        case Surface.ROTATION_270:
+          orientation = LANDSCAPE_RIGHT;
```
good catch :O

quinlanj commented:
I have implemented all the feedback in this PR. I have also split up the files of the ScreenOrientation module into different PR's. Please leave any future comments in those PR's
